[Snackbar] Create explicit singleton

romoore opened the pull request:
Creating an explicit singleton for the MDCSnackbarManager so that clients can
use their own instances (perhaps one per screen in the future) or configure
instances for testing.

Unblocks internal issue b/77900596

Commit:
[Snackbar] Create explicit singleton

Creating an explicit singleton for the MDCSnackbarManager so that clients can
use their own instances (perhaps one per screen in the future) or configure
instances for testing.

material-automation commented:
## API diff detected the following changes

### Snackbar

#### MDCSnackbarManager

*new* method: `-setPresentationHostView:` in `MDCSnackbarManager`

*new* method: `-suspendAllMessages` in `MDCSnackbarManager`

*new* method: `-suspendMessagesWithCategory:` in `MDCSnackbarManager`

*new* property: `messageTextColor` in `MDCSnackbarManager`

*new* property: `snackbarMessageViewShadowColor` in `MDCSnackbarManager`

*new* method: `-setBottomOffset:` in `MDCSnackbarManager`

*new* method: `-hasMessagesShowingOrQueued` in `MDCSnackbarManager`

*new* method: `-dismissAndCallCompletionBlocksWithCategory:` in `MDCSnackbarManager`

*new* property: `buttonFont` in `MDCSnackbarManager`

*new* property: `mdc_adjustsFontForContentSizeCategory` in `MDCSnackbarManager`

*new* method: `-buttonTitleColorForState:` in `MDCSnackbarManager`

*new* property: `alignment` in `MDCSnackbarManager`

*new* property: `delegate` in `MDCSnackbarManager`

*new* property: `defaultManager` in `MDCSnackbarManager`

*new* method: `-showMessage:` in `MDCSnackbarManager`

*new* property: `messageFont` in `MDCSnackbarManager`

*new* method: `-setButtonTitleColor:forState:` in `MDCSnackbarManager`

*new* method: `-resumeMessagesWithToken:` in `MDCSnackbarManager`

*new* property: `snackbarMessageViewBackgroundColor` in `MDCSnackbarManager`

*new* property: `shouldApplyStyleChangesToVisibleSnackbars` in `MDCSnackbarManager`

#### MDCSnackbarManager(LegacyAPI)

*new* category: `MDCSnackbarManager(LegacyAPI)`

*modified* class method: `+buttonTitleColorForState:` in `MDCSnackbarManager(LegacyAPI)`

| Type of change: | parent.usr |
|---|---|
| From: | `c:objc(cs)MDCSnackbarManager` |
| To: | `c:objc(cy)MDCSnackbarManager@LegacyAPI` |

*modified* class method: `+setBottomOffset:` in `MDCSnackbarManager(LegacyAPI)`

| Type of change: | parent.usr |
|---|---|
| From: | `c:objc(cs)MDCSnackbarManager` |
| To: | `c:objc(cy)MDCSnackbarManager@LegacyAPI` |

*modified* property: `mdc_adjustsFontForContentSizeCategory` in `MDCSnackbarManager(LegacyAPI)`

| Type of change: | parent.usr |
|---|---|
| From: | `c:objc(cs)MDCSnackbarManager` |
| To: | `c:objc(cy)MDCSnackbarManager@LegacyAPI` |

*modified* class method: `+hasMessagesShowingOrQueued` in `MDCSnackbarManager(LegacyAPI)`

| Type of change: | parent.usr |
|---|---|
| From: | `c:objc(cs)MDCSnackbarManager` |
| To: | `c:objc(cy)MDCSnackbarManager@LegacyAPI` |

*modified* class method: `+suspendMessagesWithCategory:` in `MDCSnackbarManager(LegacyAPI)`

| Type of change: | parent.usr |
|---|---|
| From: | `c:objc(cs)MDCSnackbarManager` |
| To: | `c:objc(cy)MDCSnackbarManager@LegacyAPI` |

*modified* property: `snackbarMessageViewBackgroundColor` in `MDCSnackbarManager(LegacyAPI)`

| Type of change: | parent.usr |
|---|---|
| From: | `c:objc(cs)MDCSnackbarManager` |
| To: | `c:objc(cy)MDCSnackbarManager@LegacyAPI` |

*modified* property: `shouldApplyStyleChangesToVisibleSnackbars` in `MDCSnackbarManager(LegacyAPI)`

| Type of change: | parent.usr |
|---|---|
| From: | `c:objc(cs)MDCSnackbarManager` |
| To: | `c:objc(cy)MDCSnackbarManager@LegacyAPI` |

*modified* property: `delegate` in `MDCSnackbarManager(LegacyAPI)`

| Type of change: | parent.usr |
|---|---|
| From: | `c:objc(cs)MDCSnackbarManager` |
| To: | `c:objc(cy)MDCSnackbarManager@LegacyAPI` |

*modified* class method: `+suspendAllMessages` in `MDCSnackbarManager(LegacyAPI)`

| Type of change: | parent.usr |
|---|---|
| From: | `c:objc(cs)MDCSnackbarManager` |
| To: | `c:objc(cy)MDCSnackbarManager@LegacyAPI` |

*modified* property: `buttonFont` in `MDCSnackbarManager(LegacyAPI)`

| Type of change: | parent.usr |
|---|---|
| From: | `c:objc(cs)MDCSnackbarManager` |
| To: | `c:objc(cy)MDCSnackbarManager@LegacyAPI` |

*modified* class method: `+setPresentationHostView:` in `MDCSnackbarManager(LegacyAPI)`

| Type of change: | parent.usr |
|---|---|
| From: | `c:objc(cs)MDCSnackbarManager` |
| To: | `c:objc(cy)MDCSnackbarManager@LegacyAPI` |

*modified* class method: `+dismissAndCallCompletionBlocksWithCategory:` in `MDCSnackbarManager(LegacyAPI)`

| Type of change: | parent.usr |
|---|---|
| From: | `c:objc(cs)MDCSnackbarManager` |
| To: | `c:objc(cy)MDCSnackbarManager@LegacyAPI` |

*modified* class method: `+setButtonTitleColor:forState:` in `MDCSnackbarManager(LegacyAPI)`

| Type of change: | parent.usr |
|---|---|
| From: | `c:objc(cs)MDCSnackbarManager` |
| To: | `c:objc(cy)MDCSnackbarManager@LegacyAPI` |

*modified* property: `alignment` in `MDCSnackbarManager(LegacyAPI)`

| Type of change: | parent.usr |
|---|---|
| From: | `c:objc(cs)MDCSnackbarManager` |
| To: | `c:objc(cy)MDCSnackbarManager@LegacyAPI` |

*modified* property: `snackbarMessageViewShadowColor` in `MDCSnackbarManager(LegacyAPI)`

| Type of change: | parent.usr |
|---|---|
| From: | `c:objc(cs)MDCSnackbarManager` |
| To: | `c:objc(cy)MDCSnackbarManager@LegacyAPI` |

*modified* class method: `+showMessage:` in `MDCSnackbarManager(LegacyAPI)`

| Type of change: | parent.usr |
|---|---|
| From: | `c:objc(cs)MDCSnackbarManager` |
| To: | `c:objc(cy)MDCSnackbarManager@LegacyAPI` |

*modified* property: `messageTextColor` in `MDCSnackbarManager(LegacyAPI)`

| Type of change: | parent.usr |
|---|---|
| From: | `c:objc(cs)MDCSnackbarManager` |
| To: | `c:objc(cy)MDCSnackbarManager@LegacyAPI` |

*modified* class method: `+resumeMessagesWithToken:` in `MDCSnackbarManager(LegacyAPI)`

| Type of change: | parent.usr |
|---|---|
| From: | `c:objc(cs)MDCSnackbarManager` |
| To: | `c:objc(cy)MDCSnackbarManager@LegacyAPI` |

*modified* property: `messageFont` in `MDCSnackbarManager(LegacyAPI)`

| Type of change: | parent.usr |
|---|---|
| From: | `c:objc(cs)MDCSnackbarManager` |
| To: | `c:objc(cy)MDCSnackbarManager@LegacyAPI` |

<!-- identifier: apidiff -->

Commit:
Restore theming.

jverkoey reviewed (dismissed):
Overall approach looks good. We’ll need to put all of the existing APIs through the deprecation pipeline.

Commit:
Merge branch 'develop' into snackbar-manager-shared

Commit:
Restore class properties and methods

To avoid delaying this change, we can bind the old class-based API to
the `defaultManager` instance and allow clients to move gracefully
to the new explicit singleton API.

Commit:
Revert all use of `defaultManager` in examples

To verify that our API hasn't changed, I'm reverting the examples use of
the new API. We can migrate examples to the new API in a follow-up PR.

Commit:
Revert Snackbar themer changes

Let's keep to the old API for theming in this PR to narrow the scope.

Commit:
Fixing some formatting

Commit:
Revert changes to typography themer

Commit:
Removing unit tests

Although they were the motivation for this change, we shouldn't
introduce these tests just yet.

Commit:
Fixing header

Commit:
Restoring initWithFrame:

Without it, the expected (but not designated) initializer is never
called.

Commit:
NULL -> nil

Commit:
Fix storage for instances

Adding unit tests to ensure instances are unique and don't share
storage.

Commit:
Merge branch 'develop' into snackbar-manager-shared

jverkoey commented on components/Snackbar/src/MDCSnackbarManager.m:
```diff
@@ -498,19 +491,49 @@ - (void)removeSuspensionIdentifierMainThread:(NSUUID *)identifier forCategory:(N
+  dispatch_once(&onceToken, ^{
+    kDefaultManager = [[MDCSnackbarManager alloc] init];
+  });
+  return kDefaultManager;
```
[style] We typically reserve `k` for file-level constants. Consider dropping the `k` here.

romoore replied:
Sure, how do we mark static variables that are scoped to a method/function?

romoore replied:
Also, I found that 3/4 of our uses of `dispatch_once` use the `k` prefix for the static variable being initialized. I was previously told not to use `s_` for statics (we prefer `k` for `static const`s).

* [MDCTypographySchemeFontListExampleViewController.m](https://github.com/material-components/material-components-ios/blob/develop/components/schemes/Typography/examples/MDCTypographySchemeFontListExampleViewController.m) (authored by me)
* [AnimationTimingExampleSupplemental.m](https://github.com/material-components/material-components-ios/blob/develop/components/AnimationTiming/examples/supplemental/AnimationTimingExampleSupplemental.m) (authored by you)

jverkoey replied:
We typically just name it like a local variable, so `defaultManager` in this case.

romoore replied:
:( Sounds confusing. Alright.

jverkoey commented on components/Snackbar/src/MDCSnackbarManager.m:
```diff
@@ -498,19 +491,49 @@ - (void)removeSuspensionIdentifierMainThread:(NSUUID *)identifier forCategory:(N
+}
+
+- (id<MDCSnackbarManagerDelegate>)delegate {
+
```
[stylenit] Remove this newline.

jverkoey commented on components/Snackbar/src/MDCSnackbarMessageView.m:
```diff
@@ -214,35 +214,34 @@ @implementation MDCSnackbarMessageView {
+
+- (instancetype)initWithMessage:(MDCSnackbarMessage *)message
+                 dismissHandler:(MDCSnackbarMessageDismissHandler)handler
+                snackbarManager:(MDCSnackbarManager *__weak)manager {
```
I haven't seen this pattern of marking a parameter method as `weak`; what's the advantage of doing so?

romoore replied:
I dunno, this was Xcode's autocomplete. I usually defer to the build tools to let me know the correct annotations of a definition once I've written the declaration.

jverkoey commented on components/Snackbar/src/private/MDCSnackbarMessageViewInternal.h:
```diff
@@ -62,7 +63,8 @@ typedef void (^MDCSnackbarMessageDismissHandler)(BOOL userInitiated,
 - (_Nonnull instancetype)initWithMessage:(MDCSnackbarMessage *_Nullable)message
-                          dismissHandler:(MDCSnackbarMessageDismissHandler _Nullable)handler;
+                          dismissHandler:(MDCSnackbarMessageDismissHandler _Nullable)handler
+                         snackbarManager:(MDCSnackbarManager *__weak)manager;
```
I'm assuming you may have meant to add a nullability annotation here instead of `weak`?

romoore replied:
Oops, meant that for the other comment thread. I did not intend to make it nullable, but this was left-over from when I retaining the manager. I didn't need it in the end, so I'll remove the annotation.

jverkoey replied:
I believe Xcode auto-completed this because of the header API (that I'm assuming you wrote?). As best I can tell the `__weak` is a new addition to this code and is not something we typically annotate in our APIs anywhere.

jverkoey replied:
Holy cats this comment thread got all jumbled. My reply was meant to reply to the other thread also haha.

But cool cool!

romoore replied:
I'm surprised that we don't annotate weakly-retained parameters as such. I would think we like to be explicit about when we're not retaining something passed to an initializer.

jverkoey commented on components/Snackbar/src/ColorThemer/MDCSnackbarColorThemer.m:
```diff
@@ -22,12 +22,9 @@ + (void)applySemanticColorScheme:(nonnull id<MDCColorScheming>)colorScheme {
-                                 forState:UIControlStateHighlighted];
+  UIColor *buttonTitleColor = [colorScheme.surfaceColor colorWithAlphaComponent:0.6f];
+  [MDCSnackbarManager setButtonTitleColor:buttonTitleColor forState:UIControlStateNormal];
+  [MDCSnackbarManager setButtonTitleColor:buttonTitleColor forState:UIControlStateHighlighted];
```
Consider reverting this file - looks like it's just a stylistic change.

romoore replied:
It was a result of formatting. I can revert it.

jverkoey commented on components/Snackbar/examples/SnackbarSimpleExample.m:
```diff
@@ -27,6 +27,13 @@ @implementation SnackbarSimpleExample {
+  if (!self.typographyScheme) {
+    self.typographyScheme = [[MDCTypographyScheme alloc] init];
+  }
+  
```
Consider breaking this out to a separate change.

romoore replied:
Reverted.

jverkoey commented on components/Snackbar/src/MDCSnackbarManager.h:
```diff
@@ -51,6 +51,11 @@
 @interface MDCSnackbarManager : NSObject
 
+/**
+ An instance of MDCSnackbarManager bound to the main screen.
```
Is the "bound to the main screen" statement relevant to a client? Can they change which screen the manager is bound to?

romoore replied:
Not yet, but I think there's a good use case in the future for clients to be able to send Snackbar messages to a second (or third) screen. In this case, I'm just explicitly calling-out where this Snackbar manager will send its messages.

romoore replied:
I looked a bit deeper, and it doesn't even default to the main screen of an application. It searches for any MDCOverlayWindow (hidden or not) and uses the first one it finds. If it can't, it tries the app delegate's window, and if that doesn't exist, picks the key window.

jverkoey reviewed (approved):
Overall looks good to me. All feedback is minor, with some requests for moving changes out to separate PRs because this PR has a moderate rollback risk.

yarneo reviewed (changes requested):
blocking so I can review this PR before it gets merged

Commit:
Feedback nits

Commit:
Reverting fixes to example

yarneo commented on components/Snackbar/src/MDCSnackbarMessageView.m:
```diff
@@ -214,35 +214,34 @@ @implementation MDCSnackbarMessageView {
-                 dismissHandler:(MDCSnackbarMessageDismissHandler)handler {
-  self = [super init];
-  if (self) {
+        manager.mdc_adjustsFontForContentSizeCategory;
```
indentation

romoore replied:
Could you please clarify? The indentation (4 spaces) looks correct for a continued statement.

yarneo replied:
it's a continued statement from `if (self) {` ?
meaning you wanted it to be `if (self) { manager.mdc_adjustsFontForContentSizeCategory;` but it was too long to fit in one line?

yarneo replied:
oh shoot, i didn't see that line back there, my bad. ignore

randallli commented on components/Snackbar/src/MDCSnackbarMessageView.m:
```diff
@@ -214,35 +214,34 @@ @implementation MDCSnackbarMessageView {
-}
-
-- (instancetype)initWithMessage:(MDCSnackbarMessage *)message
-                 dismissHandler:(MDCSnackbarMessageDismissHandler)handler {
```
Do we need to deprecate this? or can we just delete?

romoore replied:
This initializer is declared in [the private "internal" class-continuation category](https://github.com/material-components/material-components-ios/blob/develop/components/Snackbar/src/private/MDCSnackbarMessageViewInternal.h#L64), so I would expect it's OK for us to change/remove it without notice.

yarneo commented on components/Snackbar/src/MDCSnackbarManager.m:
```diff
@@ -498,19 +491,49 @@ - (void)removeSuspensionIdentifierMainThread:(NSUUID *)identifier forCategory:(N
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    _internalManager = [[MDCSnackbarManagerInternal alloc] initWithSnackbarManager:self];
```
does this need to pass `self`? the logic is a bit hard to follow in the PR, so just verifying this is intended.

jverkoey replied:
Was also curious about this. Had wondered if a delegate pattern might be more clear here as well, but couldn't convince myself that it would substantially improve the quality of the implementation.

@romoore I'm assuming we'll be able to remove this cyclical reference after we remove the old static APIs?

yarneo reviewed (approved):
overall looks good to me, please resolve @jverkoey and my comments, and have CI pass before merging

randallli commented on components/Snackbar/src/MDCSnackbarManager.m:
```diff
@@ -46,6 +46,12 @@ static BOOL UIViewHasFocusedAccessibilityElement(UIView *view) {
 
+/**
+ The instance of MDCSnackbarManager that "owns" this internal manager.  Used to get theming
+ properties. Can be refactored away in the future.
```
what can be refactored away? This property? or this whole class?

romoore replied:
It looks as though the entire "internal" class can be merged into `MDCSnackbarManager`. I believe it only ever existed to avoid making the singleton explicit in the first place.

randallli commented on components/Snackbar/src/MDCSnackbarManager.m:
```diff
@@ -46,6 +46,12 @@ static BOOL UIViewHasFocusedAccessibilityElement(UIView *view) {
  */
 @interface MDCSnackbarManagerInternal : NSObject
```
I think this class needs a better explanation for what its purpose is. Im struggling to understand what its being used for.

romoore replied:
????

Commit:
Merge branch 'develop' into snackbar-manager-shared

Commit:
Fix nullability annotations

romoore commented:
I'm not seeing these failures locally, but here's where kokoro is choking:

```
 C++ compilation of rule '//components/Snackbar:unit_test_sources' failed (Exit 1)
components/Snackbar/tests/unit/MDCSnackbarManagerInstanceTests.m:104:60: error: no setter method 'setMdc_adjustsFontForContentSizeCategory:' for assignment to property
  MDCSnackbarManager.mdc_adjustsFontForContentSizeCategory = YES;
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~
1 error generated.
```

Commit:
Merge branch 'develop' into snackbar-manager-shared

romoore commented:
This is not failing locally if I recreate an `ios_unit_test` target in the BUILD file:

```bash
rsmoore:material-components-ios rsmoore$ ./.kokoro --target components/Snackbar:unit_tests_simple
~/git/material-components-ios/.kokoro-ios-runner ~/git/material-components-ios
HEAD is now at e1516d0 Merge branch 'release-candidate' into stable
~/git/material-components-ios
Running bazel builds...
bazel.sh version v4.3.0
🛠️  components/Snackbar:unit_tests_simple with Xcode 9.4.1...
INFO: Starting clean.
DEBUG: /private/var/tmp/_bazel_rsmoore/9239405992f16fcf43677963a12bd325/external/build_bazel_rules_apple/apple/bundling/bundler.bzl:245:5: The target //components/Snackbar:unit_tests_simple_test_bundle does not specify its minimum OS version, so 8.0 from the --ios_minimum_os setting will be used. Please set one for this target specifically by using the minimum_os_version attribute (for example, 'minimum_os_version = "9.0"').
INFO: Analysed target //components/Snackbar:unit_tests_simple (54 packages loaded).
INFO: Found 1 test target...
INFO: From Linking components/AnimationTiming/libAnimationTiming.a:
warning: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: warning for library: bazel-out/ios-x86_64-min8.0-applebin_ios-ios_x86_64-fastbuild/bin/components/AnimationTiming/libAnimationTiming.a the table of contents is empty (no object file members in the library define global symbols)
INFO: From Linking components/private/Application/libApplication.a:
warning: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: warning for library: bazel-out/ios-x86_64-min8.0-applebin_ios-ios_x86_64-fastbuild/bin/components/private/Application/libApplication.a the table of contents is empty (no object file members in the library define global symbols)
INFO: From Linking components/AnimationTiming/libAnimationTiming.a:
warning: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: warning for library: bazel-out/ios-i386-min8.0-applebin_ios-ios_i386-fastbuild/bin/components/AnimationTiming/libAnimationTiming.a the table of contents is empty (no object file members in the library define global symbols)
INFO: From Linking components/private/Application/libApplication.a:
warning: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: warning for library: bazel-out/ios-i386-min8.0-applebin_ios-ios_i386-fastbuild/bin/components/private/Application/libApplication.a the table of contents is empty (no object file members in the library define global symbols)
INFO: From Processing and signing IosTest: unit_tests_simple:
bazel-out/darwin-fastbuild/bin/components/Snackbar/unit_tests_simple_test_bundle.archive-root/unit_tests_simple.xctest/Frameworks/libswiftCore.dylib: replacing existing signature
bazel-out/darwin-fastbuild/bin/components/Snackbar/unit_tests_simple_test_bundle.archive-root/unit_tests_simple.xctest/Frameworks/libswiftCoreAudio.dylib: replacing existing signature
bazel-out/darwin-fastbuild/bin/components/Snackbar/unit_tests_simple_test_bundle.archive-root/unit_tests_simple.xctest/Frameworks/libswiftCoreFoundation.dylib: replacing existing signature
bazel-out/darwin-fastbuild/bin/components/Snackbar/unit_tests_simple_test_bundle.archive-root/unit_tests_simple.xctest/Frameworks/libswiftCoreGraphics.dylib: replacing existing signature
bazel-out/darwin-fastbuild/bin/components/Snackbar/unit_tests_simple_test_bundle.archive-root/unit_tests_simple.xctest/Frameworks/libswiftCoreImage.dylib: replacing existing signature
bazel-out/darwin-fastbuild/bin/components/Snackbar/unit_tests_simple_test_bundle.archive-root/unit_tests_simple.xctest/Frameworks/libswiftCoreMedia.dylib: replacing existing signature
bazel-out/darwin-fastbuild/bin/components/Snackbar/unit_tests_simple_test_bundle.archive-root/unit_tests_simple.xctest/Frameworks/libswiftDarwin.dylib: replacing existing signature
bazel-out/darwin-fastbuild/bin/components/Snackbar/unit_tests_simple_test_bundle.archive-root/unit_tests_simple.xctest/Frameworks/libswiftDispatch.dylib: replacing existing signature
bazel-out/darwin-fastbuild/bin/components/Snackbar/unit_tests_simple_test_bundle.archive-root/unit_tests_simple.xctest/Frameworks/libswiftFoundation.dylib: replacing existing signature
bazel-out/darwin-fastbuild/bin/components/Snackbar/unit_tests_simple_test_bundle.archive-root/unit_tests_simple.xctest/Frameworks/libswiftMetal.dylib: replacing existing signature
bazel-out/darwin-fastbuild/bin/components/Snackbar/unit_tests_simple_test_bundle.archive-root/unit_tests_simple.xctest/Frameworks/libswiftObjectiveC.dylib: replacing existing signature
bazel-out/darwin-fastbuild/bin/components/Snackbar/unit_tests_simple_test_bundle.archive-root/unit_tests_simple.xctest/Frameworks/libswiftQuartzCore.dylib: replacing existing signature
bazel-out/darwin-fastbuild/bin/components/Snackbar/unit_tests_simple_test_bundle.archive-root/unit_tests_simple.xctest/Frameworks/libswiftUIKit.dylib: replacing existing signature
bazel-out/darwin-fastbuild/bin/components/Snackbar/unit_tests_simple_test_bundle.archive-root/unit_tests_simple.xctest/Frameworks/libswiftXCTest.dylib: replacing existing signature
bazel-out/darwin-fastbuild/bin/components/Snackbar/unit_tests_simple_test_bundle.archive-root/unit_tests_simple.xctest/Frameworks/libswiftos.dylib: replacing existing signature
Target //components/Snackbar:unit_tests_simple up-to-date:
  bazel-bin/components/Snackbar/unit_tests_simple.zip
  bazel-bin/components/Snackbar/unit_tests_simple
INFO: Elapsed time: 49.459s, Critical Path: 12.68s
INFO: Build completed successfully, 298 total actions
//components/Snackbar:unit_tests_simple                                  PASSED in 4.7s

Executed 1 out of 1 test: 1 test passes.
There were tests whose specified size is too big. Use the --test_verbose_timeout_warnings command line option to see which ones these are.
Success!
rsmoore:material-components-ios rsmoore$
```

romoore commented:
Assigning to Jeff to help figure out why the kokoro build system isn't parsing headers correctly.

jverkoey commented:
Did you try `mdc_setAdjustsFontForContentSizeCategory`? I'm guessing there's something to do here with the intersection of a class property and a custom setter.

romoore commented:
That's the property I'm currently using.  This was previously a class property and it has no problems compiling locally using Xcode and bazel.

jverkoey commented:
I understand, but you presently have written `mdc_adjustsFontForContentSizeCategory`, I was curious if `mdc_setAdjustsFontForContentSizeCategory` was possible in this context?

jverkoey commented:
The error gives a hint here:

> no setter method 'setMdc_adjustsFontForContentSizeCategory:' for assignment to property

It seems to be mistakenly attempting to infer a selector that isn't the one we've explicitly defined (`mdc_setAdjustsFontForContentSizeCategory`).

romoore commented:
I understand the error - I've been working on this problem for a few days. I'm pulling in your help to try and identify how to get bazel/kokoro working.  If it's not at parity with local bazel builds then we may have a significant error in our configuration.

jverkoey commented:
To clarify what I'm trying to understand, did you already try applying the following change to your PR?

At line 104:

```
-  MDCSnackbarManager.mdc_adjustsFontForContentSizeCategory = YES;
+  [MDCSnackbarManager mdc_setAdjustsFontForContentSizeCategory:YES];
```

(Edit by @romoore to fix syntax)

romoore commented:
I did not switch to the setter since there is no failure locally via either normal Xcode builds nor with bazel. I'm trying to diagnose the problem on the back-end with kokoro so we can root cause the issue.

jverkoey commented:
I understand that you want to find an ideal solution.

Can you help me identify whether there's any way to unblock you by trying that setter?

jverkoey commented:
To be abundantly clear: I have no subtext in my question (I'm not implying we won't try to find the root cause, that you haven't tried this, that you don't understand the error, or that you don't intend to find a good solution), I simply want to understand whether that change fixes the build failure.

Commit:
Use setter

romoore commented:
Thanks, @jverkoey. I'm not trying to get around the build failures (I could have done that last week), I'm trying to figure out what's gone wrong on kokoro.  

Since we depend on it for verifying our builds and it seems to have problems parsing pretty typical Objective-C code I pulled you in to help diagnose the problem. I believe you're our point of contact with the kokoro/bazel team (or you at least have the most experience with them), so I was hoping you could assist me in debugging.

jverkoey commented:
I'm sorry; some clarity seems to be getting lost in this textual medium:

I understand and recognize that this is a problem with kokoro/bazel and that we're encountering some deltas between local builds and CI builds. This is a concern and we do need to look into it.

Separately, you have a fairly large PR that appears ready to land except for this particular issue.

**These are two separate problems**. I want you to be able to proceed with your work by unblocking the build failure on this PR because it should not be blocked by the separate CI issue.

Separately, we can explore what's causing this discrepancy, but this work does not have to block this PR from landing.

romoore commented:
Agreed, and I apologize for not being clearer in my comment about assigning you. I'm hoping you can assist me in figuring out the problems with kokoro that are being exposed by this PR.

Commit:
Restore property assignment

jverkoey commented:
In the interest of getting this PR landed, can we use the setter that does pass the build?

romoore commented:
No, I don't believe we should be merging code that fails to work correctly on our supported build platforms. I don't want to have to revert this change because a bunch of internal clients' code breaks as a result of this problem.

ianegordon commented on components/Snackbar/src/MDCSnackbarManager.h:
```diff
@@ -51,6 +51,11 @@
+/**
+ An instance of MDCSnackbarManager.
+ */
+@property(class, nonnull, nonatomic, readonly, strong) MDCSnackbarManager *defaultManager;
```
Will the prevent us from ever presenting a snackbar on a screen that isn't mainScreen?

romoore replied:
Nope, it would in fact permit greater flexibility in binding snackbar managers to windows.

jverkoey commented:
I was able to reproduce the build failures locally with Xcode 9.1, but not on Xcode 10.0 beta 4.

As best I can tell here, this appears to be a bug in Objective-C's support for class properties on older Xcode versions. It appears that a class property with a custom setter will correctly enforce the implementation of the custom setter, but when used in client code Objective-C will attempt to look up the default setter instead.

I see a few potential paths forward here:

1. We get rid of the custom setter for the class property. This allows clients to use property notation as expected, but we have a weird looking setter.
2. We drop the `mdc_` prefix. This removes the need for a custom setter.

Will leave the decision up to you @romoore, but given that this is fixed in Xcode 10 I'm not confident that the radar will ever be resolved on the older Xcode versions.

Patch of the proposed changes for option 1:

```
diff --git a/components/Snackbar/examples/SnackbarSimpleExample.m b/components/Snackbar/examples/SnackbarSimpleExample.m
index d9ebad67..3043c116 100644
--- a/components/Snackbar/examples/SnackbarSimpleExample.m
+++ b/components/Snackbar/examples/SnackbarSimpleExample.m
@@ -68,7 +68,7 @@
   } else {
     [self.navigationItem.rightBarButtonItems.lastObject setTitle:@"DT Off"];
   }
-  [MDCSnackbarManager mdc_setAdjustsFontForContentSizeCategory:_dynamicType];
+  MDCSnackbarManager.mdc_adjustsFontForContentSizeCategory = _dynamicType;
 }
 
 #pragma mark - Event Handling
diff --git a/components/Snackbar/src/MDCSnackbarManager.h b/components/Snackbar/src/MDCSnackbarManager.h
index 2d2d720b..bffe7f87 100644
--- a/components/Snackbar/src/MDCSnackbarManager.h
+++ b/components/Snackbar/src/MDCSnackbarManager.h
@@ -334,8 +334,7 @@
 /**
  Bound to @c mdc_adjustsFontForContentSizeCategory on the @c defaultManager instance.
  */
-@property(class, nonatomic, readwrite, setter=mdc_setAdjustsFontForContentSizeCategory:)
-    BOOL mdc_adjustsFontForContentSizeCategory;
+@property(class, nonatomic, readwrite) BOOL mdc_adjustsFontForContentSizeCategory;
 
 /**
  Bound to @c delegate on the @c defaultManager instance.
diff --git a/components/Snackbar/src/MDCSnackbarManager.m b/components/Snackbar/src/MDCSnackbarManager.m
index 7b1998a1..357243e9 100644
--- a/components/Snackbar/src/MDCSnackbarManager.m
+++ b/components/Snackbar/src/MDCSnackbarManager.m
@@ -867,7 +867,7 @@ static NSString *const kAllMessagesCategory = @"$$___ALL_MESSAGES___$$";
   return MDCSnackbarManager.defaultManager.mdc_adjustsFontForContentSizeCategory;
 }
 
-+ (void)mdc_setAdjustsFontForContentSizeCategory:(BOOL)mdc_adjustsFontForContentSizeCategory {
++ (void)setMdc_adjustsFontForContentSizeCategory:(BOOL)mdc_adjustsFontForContentSizeCategory {
   [MDCSnackbarManager.defaultManager
       mdc_setAdjustsFontForContentSizeCategory:mdc_adjustsFontForContentSizeCategory];
 }
```

Commit:
Merge branch 'develop' into snackbar-manager-shared

Commit:
Move legacy API out of a category

It seems older Xcode versions (e.g., 9.1) cannot handle custom setters
on a class property defined in a category. Switching back to the main
interface declaration to see if that fixes the problems.

romoore commented:
Ok. Since kokoro is failing with 9.2 and locally I'm passing with 9.4.1, it seems it was fixed somewhere between those two versions. I'll restore the original properties to the main class interface declaration and see if that fixes things.

romoore commented:
Looks like things are still hosed.  Since there are no existing internal references to this property for internal clients (or our own code), I suspect that this was a problem from the beginning. Since there was no code that used it, it never appeared as an error. I'll switch to using the setter in the unit tests and then merge.

Commit:
Use explicit setter for class property

It appears that Xcode simply cannot handle custom setters on class
properties at all (at least for earlier versions). Since none of our
code previously invoked the property setter, we didn't know it was
failing.

I don't want this to be a breaking change, so we'll keep the current
API. The unit test that is setting the property will use the explicit
setter instead.

romoore commented:
Please merge this if it passes CI.

jverkoey commented:
This PR was reverted in v59.1.0 and will need to be rolled forward again in the next release.
